internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const (
+	requestStart = `
+{
+  "streams": [
+    {
+      "stream": {
+        "label": "pebble"
+      },
+      "values": [`
+
+	requestEnd = `
+      ]
+    }
+  ]
+}`
+)
+
+// lokiClient implements logClient
+// We send a new request to Loki when either:
+//   - we have maxMessages ready to send; or
+//   - delay has passed since the last request.
+type lokiClient struct {
+	lokiURL     string
+	maxMessages int
+	delay       time.Duration
+
+	// buffered request so far
+	buffer *bytes.Buffer
+	// how many messages have been written so far
+	messages int
+	// notify when buffer fills up
+	writeCh chan struct{}
+	// notify when deadline elapses
+	timeCh <-chan time.Time
+}
+
+func newLokiClient(lokiURL string, maxMessages int, delay time.Duration) *lokiClient {
+	c := &lokiClient{
+		lokiURL:     lokiURL,
+		maxMessages: maxMessages,
+		delay:       delay,
+	}
+	c.Start()
+	return c
+}
+
+func (c *lokiClient) Start() {
+	c.reset()
+	go c.run()
+}
+
+func (c *lokiClient) reset() {
```

Comment:
Reset should only work when there is no go routine running. This one is going to be tricky to deal with I think.